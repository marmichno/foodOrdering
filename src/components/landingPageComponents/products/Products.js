import {useState, useEffect} from 'react';
import {productGetByIdRequest} from '../../../requests/productGetByIdRequest';
import {useDispatch, useSelector} from 'react-redux';
import {addProduct} from '../../../actions/index'
import {motion} from 'framer-motion';

const containerVariants = {
  hidden: {
      x: 2000
  },
  visible: {
      x: 0,
      transition: {
        type: 'spring',
        delay: 0.5,
        mass: 1,
        damping: 12
      }
  },
  exit: {
      x:'100vw',
      transition: {ease: 'easeInOut', duration:0.5}
  }
}

export const Products = () =>{

    const dispatch = useDispatch();
    const productGroupId = useSelector(state => state.orderGroupIdReducer);
    const [products, setProducts] = useState([]);


    useEffect(() => {
      getProducts(productGroupId);
    },[])

    useEffect(() =>{
      getProducts();
    },[productGroupId]);

    const getProducts = async () => {
      if(productGroupId !== null){
        const response = await productGetByIdRequest(productGroupId);
        if(response !== undefined){
          setProducts(response);
        }
      }
    }

      const addToCart = (e) => {
        const price = e.target.dataset.price;
        const name = e.target.dataset.name;

        const productData = {
            price,
            name,
            quantity: 1
        }

        dispatch(addProduct(productData));
    }

      return(
        <motion.div className="foodMainContainer"
        variants={containerVariants}
        initial="hidden"
        animate="visible"
        exit="exit"
        >
              {products !== undefined ?
              products.map((value, id) => {
                return (
                  <div key={`foodWrapper${id}`} className="foodWrapper">
                    <div key={id} className="foodContainer">
                      <div className="productImage">
                      </div>
                      <div className="productDetails">
                        <div className="productHeader">
                          <h1>{value.productName}</h1>
                          <hr></hr>
                        </div>
                        <div className="productIngredients">
                          <p>{value.description}</p>
                          <h1>~ {value.price}</h1>
                        </div>
                        <button data-testid={`addToCartBtn${id}`} data-price={value.price} data-name={value.productName} onClick={addToCart}>Order</button>
                      </div>
                    </div>
                  </div>
                )
              })
              : null
              }
        </motion.div>
        )
}

export default Products;