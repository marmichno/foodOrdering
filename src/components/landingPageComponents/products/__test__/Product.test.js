import Products from '../Products';
import validProductsJson from '../__mocks__/validProducts.json';
import mockedAxios from 'axios';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux'; 
import {render, fireEvent, waitFor} from '@testing-library/react';
import TestRenderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";

const store = createStore(() => [], {}, applyMiddleware());

test("checks if products are visible after api call", async () => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductsJson});

    const {getAllByText} = render(<Provider store={store}><Products/></Provider>);

    await waitFor(() => {
        expect(getAllByText('philadelphia'));
        expect(getAllByText('rice,nori,salmon,philadelphia cream'));
        expect(getAllByText('~ 6.00'));
    })
});

test("check if button calls dispatch", async () => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductsJson});

    const {getByTestId} = render(<Provider store={store}><Products/></Provider>);

    store.dispatch = jest.fn();

    await waitFor(() => {
        fireEvent.click(getByTestId("addToCartBtn0"));
    })


    expect(store.dispatch).toHaveBeenCalledTimes(1);
});

test("matches snapshot", () => {
    const tree = TestRenderer.create(<Provider store={store}><Products/></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
});
