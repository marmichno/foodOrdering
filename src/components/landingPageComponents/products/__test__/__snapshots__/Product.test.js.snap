// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="foodMainContainer"
  style={
    Object {
      "transform": "translateX(2000px) translateZ(0)",
    }
  }
/>
`;
